README.md
[![Release](https://img.shields.io/github/release/giterlizzi/perl-Crypt-PBE.svg)](https://github.com/giterlizzi/perl-Crypt-PBE/releases) [![Build Status](https://travis-ci.org/giterlizzi/perl-Crypt-PBE.svg)](https://travis-ci.org/giterlizzi/perl-Crypt-PBE) [![License](https://img.shields.io/github/license/giterlizzi/perl-Crypt-PBE.svg)](https://github.com/giterlizzi/perl-Crypt-PBE) [![Starts](https://img.shields.io/github/stars/giterlizzi/perl-Crypt-PBE.svg)](https://github.com/giterlizzi/perl-Crypt-PBE) [![Forks](https://img.shields.io/github/forks/giterlizzi/perl-Crypt-PBE.svg)](https://github.com/giterlizzi/perl-Crypt-PBE) [![Issues](https://img.shields.io/github/issues/giterlizzi/perl-Crypt-PBE.svg)](https://github.com/giterlizzi/perl-Crypt-PBE/issues) [![Coverage Status](https://coveralls.io/repos/github/giterlizzi/perl-Crypt-PBE/badge.svg)](https://coveralls.io/github/giterlizzi/perl-Crypt-PBE)

# Crypt::PBE - Perl extension for PKCS #5 Password-Based Encryption

## Synopsis

```.pl
use Crypt::PBE qw(:jce);

my $pbe = PBEWithMD5AndDES('mypassword');

my $encrypted = $pbe->encrypt('secret'); # Base64 encrypted data

print $pbe->decrypt($encrypted);
```

## Install

To install `Crypt::PBE` distribution, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

## Copyright

 - Copyright 2020 © Giuseppe Di Terlizzi


Changes
0.100 2020/07/15
    - initial release



Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Crypt::PBE',
    AUTHOR             => q{Giuseppe Di Terlizzi <gdt@cpan.org>},
    VERSION_FROM       => 'lib/Crypt/PBE.pm',
    ABSTRACT_FROM      => 'lib/Crypt/PBE.pm',
    LICENSE            => 'artistic_2',
    EXE_FILES          => ['bin/pkcs5-tool'],
    MIN_PERL_VERSION   => 5.008,
    PL_FILES           => {},
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    TEST_REQUIRES => {
        'Test::More' => '0',
    },
    PREREQ_PM => {
        'Crypt::CBC'  => '0',
        'Crypt::DES'  => '0',
        'Digest::MD2' => '0',
        'Digest::MD5' => '0',
        'Digest::SHA' => '0',
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        'resources' => {
            repository => {
                type => 'git',
                url  => 'git://github.com/giterlizzi/perl-Crypt-PBE',
                web  => 'https://github.com/giterlizzi/perl-Crypt-PBE'
            },
        }
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES    => 'Crypt-PBE-*' },
);


lib/Crypt/PBE/PBES1.pm
package Crypt::PBE::PBES1;

use strict;
use warnings;

use Carp;
use Crypt::CBC;

use Crypt::PBE::PBKDF1;

use Exporter qw(import);

our $VERSION = '0.100';

use constant ENCRYPTION => { 'des' => 'Crypt::DES', };

sub new {

    my ( $class, %params ) = @_;

    my $self = {
        password   => $params{password}   || croak('Specify password'),
        count      => $params{count}      || 1_000,
        hash       => $params{hash}       || croak('Specify hash digest algorithm'),
        encryption => $params{encryption} || 'des',                                    # TODO add support for RC2
        dk_len     => 0,
    };

    my $dk_len = 20;
    $dk_len = 16 if ( $self->{hash} =~ '/md(2|5)/' );

    $self->{dk_len} = $dk_len;

    return bless $self, $class;

}

sub encrypt {

    my ( $self, $data ) = @_;

    my $salt = Crypt::CBC->random_bytes(8);
    my $DK   = pbkdf1(
        hash     => $self->{hash},
        password => $self->{password},
        salt     => $salt,
        count    => $self->{count},
        dk_len   => $self->{dl_len}
    );

    my $key = substr( $DK, 0, 8 );
    my $iv  = substr( $DK, 8, 8 );

    my $crypt = Crypt::CBC->new(
        -key         => $key,
        -iv          => $iv,
        -literal_key => 1,
        -header      => 'none',
        -cipher      => 'Crypt::DES',
    );

    my @result = ( $salt, $crypt->encrypt($data) );

    return wantarray ? @result : join( '', @result );

}

sub decrypt {

    my ( $self, $salt, $encrypted ) = @_;

    if ( !$encrypted ) {
        my $data = $salt;
        $salt      = substr( $data, 0, 8 );
        $encrypted = substr( $data, 8 );
    }

    my $DK = pbkdf1(
        hash     => $self->{hash},
        password => $self->{password},
        salt     => $salt,
        count    => $self->{count},
        dk_len   => $self->{dl_len}
    );

    my $key = substr( $DK, 0, 8 );
    my $iv  = substr( $DK, 8, 8 );

    my $ciper = Crypt::CBC->new(
        -key         => $key,
        -iv          => $iv,
        -literal_key => 1,
        -header      => 'none',
        -cipher      => 'Crypt::DES',
    );

    my $decrypted = $ciper->decrypt($encrypted);

    return $decrypted;

}

1;

=head1 NAME

Crypt::PBE::PBES1 - Perl extension for PKCS #5 Password-Based Encryption Schema 1 (PBES1)

=head1 SYNOPSIS

    use Crypt::PBE::PBES1;

    my $pbes1 = Crypt::PBE::PBES1->new(
        'hash'       => 'md5',
        'encryption' => 'des',
        'password'   => 'mypassword'
    );

    my $encrypted = $pbes1->encrypt('secret');
    say $pbes1->decrypt($encrypted); # secret


=head1 DESCRIPTION

PBES1 combines the PBKDF1 function with an underlying block cipher, which shall
be either DES or RC2 in cipher block chaining (CBC) mode.

PBES1 is recommended only for compatibility with existing applications, since it
supports only two underlying encryption schemes, each of which has a key size
(56 or 64 bits) that may not be large enough for some applications.


=head1 CONSTRUCTOR

=head2 Crypt::PBE::PBES1->new ( %params )

Params:

=over 4

=item * C<password> : The password to use for the derivation

=item * C<hash> : Hash algorithm ("md2", "md5 or "sha1")

=item * C<encryption> : Encryption method ("des")

=item * C<count> : The number of internal iteractions to perform for the derivation key (default "1_000")

=back


=head1 METHODS

=head2 $pbes1->encrypt( $message )

Perform the encryption and return the encrypted message in binary format.

You can encode the encrypted message in Base64 using L<MIME::Base64>:

    $b64_encrypted = encode_base64 $pbes1->encrypt('secret');


=head2 $pbes1->decrypt( $data )

Perform the decryption and return the original message.

    $decrypted = $pbes1->decrypt($encrypted);

You can decode the Base64 encrypted message using L<MIME::Base64>:

    $decrypted = $pbes1->decrypt(decode_base64 $b64_encrypted);


=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/giterlizzi/perl-Crypt-PBE/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/giterlizzi/perl-Crypt-PBE>

    git clone https://github.com/giterlizzi/perl-Crypt-PBE.git


=head1 AUTHOR

=over 4

=item * Giuseppe Di Terlizzi <gdt@cpan.org>

=back


=head1 SEE ALSO

=over 4

=item L<Crypt::PBE::PBKDF1>

=item L<Crypt::PBE::PBES2>

=item [RFC2898] PKCS #5: Password-Based Cryptography Specification Version 2.0 (L<https://tools.ietf.org/html/rfc2898>)

=item [RFC8018] PKCS #5: Password-Based Cryptography Specification Version 2.1 (L<https://tools.ietf.org/html/rfc8018>)

=back


=head1 LICENSE AND COPYRIGHT

This software is copyright (c) 2020 by Giuseppe Di Terlizzi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Crypt/PBE/PBES2.pm
package Crypt::PBE::PBES2;

use strict;
use warnings;

use Carp;
use Crypt::CBC;

use Crypt::PBE::PBKDF2;

use Exporter qw(import);

our $VERSION = '0.100';

use constant KEY_SIZE => {
    'aes-128' => 16,
    'aes-192' => 24,
    'aes-256' => 32,
};

use constant ENCRYPTION => {
    'aes-128' => 'Crypt::OpenSSL::AES',
    'aes-192' => 'Crypt::OpenSSL::AES',
    'aes-256' => 'Crypt::OpenSSL::AES',
};

use constant HMAC => qw( hmac-sha1 hmac-224 hmac-sha256 hmac-sha384 hmac-sha512 );

sub new {

    my ( $class, %params ) = @_;

    my $self = {
        password   => $params{password}   || croak('Specify password'),
        hmac       => $params{hmac}       || croak('Specify HMAC digest algorithm'),
        encryption => $params{encryption} || croak('Specify encryption method'),
        count      => $params{count}      || 1_000,
        dk_len     => 0,
    };

    my $dk_len = KEY_SIZE->{ $params{encryption} };

    if ( !$dk_len ) {
        croak('Unknown encryption method');
    }

    $self->{dk_len} = $dk_len;

    return bless $self, $class;

}

sub encrypt {

    my ( $self, $data ) = @_;

    my $salt = Crypt::CBC->random_bytes(16);
    my $iv   = Crypt::CBC->random_bytes(16);

    my $key = pbkdf2(
        prf      => $self->{hmac},
        password => $self->{password},
        salt     => $salt,
        dk_len   => $self->{dk_len},
        count    => $self->{count}
    );

    my $cipher = Crypt::CBC->new(
        -key         => $key,
        -keysize     => length($key),
        -iv          => $iv,
        -header      => 'none',
        -literal_key => 1,
        -cipher      => ENCRYPTION->{ $self->{encryption} }
    );

    my $encrypted = $cipher->encrypt($data);

    my @result = ( $salt, $iv, $encrypted );

    return wantarray ? @result : join( '', @result );

}

sub decrypt {

    my ( $self, $salt, $iv, $encrypted ) = @_;

    if ( !$iv && !$encrypted ) {

        my $data = $salt;

        $salt = substr( $data, 0,  16 );
        $iv   = substr( $data, 16, 16 );
        $encrypted = substr( $data, 32 );

    }

    my $key = pbkdf2(
        prf      => $self->{hmac},
        password => $self->{password},
        salt     => $salt,
        dk_len   => $self->{dk_len},
        count    => $self->{count}
    );

    my $cipher = Crypt::CBC->new(
        -key         => $key,
        -keysize     => length($key),
        -iv          => $iv,
        -header      => 'none',
        -literal_key => 1,
        -cipher      => ENCRYPTION->{ $self->{encryption} }
    );

    return $cipher->decrypt($encrypted);

}

1;

=head1 NAME

Crypt::PBE::PBES2 - Perl extension for PKCS #5 Password-Based Encryption Schema 2 (PBES2)

=head1 SYNOPSIS

    use Crypt::PBE::PBES2;

    my $pbes2 = Crypt::PBE::PBES2->new(
        'hmac'       => 'hmac-sha256',
        'encryption' => 'aes-128',
        'password'   => 'mypassword'
    );

    my $encrypted = $pbes2->encrypt('secret');
    say $pbes2->decrypt($encrypted); # secret


=head1 DESCRIPTION

PBES2 combines a password-based key derivation function, which shall be PBKDF2 
with an underlying encryption scheme.  The key length and any other parameters
for the underlying encryption scheme depend on the scheme.

PBES2 is recommended for new applications.


=head1 CONSTRUCTOR

=head2 Crypt::PBE::PBES2->new ( %params )

Params:

=over 4

=item * C<password> : The password to use for the derivation

=item * C<hmac> : HMAC digest algorithm ("hmac-sha1", "hmac-sha224", "hmac-sha256", "hmac-sha384" or "hmac-512")

=item * C<encryption> : Encryption method ("aes-128", "aes-192" or "aes-256")

=item * C<count> : The number of internal iteractions to perform for the derivation key (default "1_000")

=back


=head1 METHODS

=head2 $pbes2->encrypt( $message )

Perform the encryption and return the encrypted message in binary format.

You can encode the encrypted message in Base64 using L<MIME::Base64>:

    $b64_encrypted = encode_base64 $pbes2->encrypt('secret');


=head2 $pbes2->decrypt( $data )

Perform the decryption and return the original message.

    $decrypted = $pbes2->decrypt($encrypted);

You can decode the Base64 encrypted message using L<MIME::Base64>:

    $decrypted = $pbes2->decrypt(decode_base64 $b64_encrypted);


=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/giterlizzi/perl-Crypt-PBE/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/giterlizzi/perl-Crypt-PBE>

    git clone https://github.com/giterlizzi/perl-Crypt-PBE.git


=head1 AUTHOR

=over 4

=item * Giuseppe Di Terlizzi <gdt@cpan.org>

=back


=head1 SEE ALSO

=over 4

=item L<Crypt::PBE::PBKDF2>

=item L<Crypt::PBE::PBES1>

=item [RFC2898] PKCS #5: Password-Based Cryptography Specification Version 2.0 (L<https://tools.ietf.org/html/rfc2898>)

=item [RFC8018] PKCS #5: Password-Based Cryptography Specification Version 2.1 (L<https://tools.ietf.org/html/rfc8018>)

=back


=head1 LICENSE AND COPYRIGHT

This software is copyright (c) 2020 by Giuseppe Di Terlizzi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Crypt/PBE/PBKDF1.pm
package Crypt::PBE::PBKDF1;

use strict;
use warnings;

use Carp;

use MIME::Base64;

use Digest::MD2 qw(md2);
use Digest::MD5 qw(md5);
use Digest::SHA qw(sha1);

use Exporter qw(import);

our $VERSION = '0.100';

our @EXPORT = qw(
    pbkdf1
    pbkdf1_base64
    pbkdf1_hex
);

our @EXPORT_OK = qw(
    pbkdf1_md2
    pbkdf1_md2_base64
    pbkdf1_md2_hex

    pbkdf1_md5
    pbkdf1_md5_base64
    pbkdf1_md5_hex

    pbkdf1_sha1
    pbkdf1_sha1_base64
    pbkdf1_sha1_hex
);

sub new {

    my ( $class, %params ) = @_;

    my $password = delete $params{password} || croak 'Specify password';
    my $salt     = delete $params{salt}     || croak 'Specify salt';
    my $count    = delete $params{count}    || 1_000;
    my $hash     = delete $params{hash}     || 'sha1';
    my $dk_len   = 20;

    $dk_len = 16 if ( $hash eq 'md5' || $hash eq 'md2' );

    my $self = {
        password => $password,
        salt     => $salt,
        count    => $count,
        hash     => $hash,
        dk_len   => $dk_len
    };

    bless $self, $class;

    return $self;

}

sub hash_algorithm {
    my ($self) = @_;
    return $self->{hash};
}

sub count {
    my ($self) = @_;
    return $self->{count};
}

sub derived_key_length {
    my ($self) = @_;
    return $self->{dk_len};
}

sub derived_key {
    my ($self) = @_;
    return pbkdf1(
        password => $self->{password},
        salt     => $self->{salt},
        count    => $self->{count},
        hash     => $self->{hash},
        dk_len   => $self->{dk_len}
    );
}

sub derived_key_base64 {
    my ($self) = @_;
    return pbkdf1_base64(
        password => $self->{password},
        salt     => $self->{salt},
        count    => $self->{count},
        hash     => $self->{hash},
        dk_len   => $self->{dk_len}
    );
}

sub derived_key_hex {
    my ($self) = @_;
    return pbkdf1_hex(
        password => $self->{password},
        salt     => $self->{salt},
        count    => $self->{count},
        hash     => $self->{hash},
        dk_len   => $self->{dk_len}
    );
}

# PBKDF1 (P, S, c, dkLen)
#
#   Options:        Hash       underlying hash function
#
#   Input:          P          password, an octet string
#                   S          salt, an octet string
#                   c          iteration count, a positive integer
#                   dkLen      intended length in octets of derived key,
#                              a positive integer, at most 16 for MD2 or
#                              MD5 and 20 for SHA-1
#   Output:         DK         derived key, a dkLen-octet string

sub pbkdf1 {

    my (%params) = @_;

    my $hash = delete( $params{hash} )     || 'sha1';
    my $P    = delete( $params{password} ) || croak('Specify password');
    my $S    = delete( $params{salt} )     || croak('Specify salt');
    my $c    = delete( $params{count} )    || 1_000;
    my $dkLen = delete( $params{dk_len} );

    if ( $hash ne 'md2' && $hash ne 'md5' && $hash ne 'sha1' ) {
        croak 'unknown hash function';
    }

    if ( !$dkLen ) {
        $dkLen = 16 if ( $hash =~ /md(2|5)/ );
        $dkLen = 20 if ( $hash eq 'sha1' );
    }

    if ( ( $hash eq 'md5' && $dkLen > 16 ) || ( $hash eq 'md2' && $dkLen > 16 ) || ( $hash eq 'sha1' && $dkLen > 20 ) )
    {
        croak 'derived key too long';
    }

    my $T = $P . $S;

    no strict 'refs';    ## no critic

    for ( 1 .. $c ) {
        $T = &{$hash}($T);
    }

    my $DK = substr( $T, 0, $dkLen );

    return $DK;

}

sub pbkdf1_hex {
    return join '', unpack '(H2)*', pbkdf1(@_);
}

sub pbkdf1_base64 {
    return encode_base64 pbkdf1(@_), '';
}

for my $digest (qw/md2 md5 sha1/) {

    my $dk_len = 16;
    $dk_len = 20 if ( $digest eq 'sha1' );

    my $sub_name = 'pbkdf1_' . $digest;

    no strict 'refs';    ## no critic

    *{$sub_name} = sub {
        my (%params) = @_;
        $params{hash}   = $digest;
        $params{dk_len} = $dk_len;
        return pbkdf1(%params);
    };

    *{ $sub_name . '_base64' } = sub {
        my (%params) = @_;
        $params{hash}   = $digest;
        $params{dk_len} = $dk_len;
        return encode_base64 pbkdf1(%params), '';
    };

    *{ $sub_name . '_hex' } = sub {
        my (%params) = @_;
        $params{hash}   = $digest;
        $params{dk_len} = $dk_len;
        return join '', unpack '(H2)*', pbkdf1(%params);
    };

}

1;
__END__
=head1 NAME

Crypt::PBE::PBKDF1 - Perl extension for PKCS #5 Password-Based Key Derivation Function 1 (PBKDF1)

=head1 SYNOPSIS

    use Crypt::PBE::PBKDF1;

    # OO style

    my $pbkdf1 = Crypt::PBE::PBKDF1->new(
        password   => $password,
        salt       => $salt,
        hash       => 'sha1'
    );

    $pbkdf1->derived_key;           # Byte
    $pbkdf1->derived_key_base64     # Base64 encoded
    $pbkdf1->derived_key_hex        # Hex


    use Crypt::PBE::PBKDF1 qw(pbkdf1_md2 pbkdf1_md5_base64 pbkdf1_sha1_hex);

    # Functional style

    $derived_key = pbkdf1 ( %params );           # Byte
    $derived_key = pbkdf1_base64 ( %params );    # Base64 encoded
    $derived_key = pbkdf1_hex ( %params );       # Hex

    # Functional style helpers
    $derived_key = pbkdf1_md2 ( %params );
    $derived_key = pbkdf1_md2_hex ( %params );
    $derived_key = pbkdf1_md2_base64 ( %params );

    $derived_key = pbkdf1_md5 ( %params );
    $derived_key = pbkdf1_md5_hex ( %params );
    $derived_key = pbkdf1_md5_base64 ( %params );

    $derived_key = pbkdf1_sha1 ( %params );
    $derived_key = pbkdf1_sha1_hex ( %params );
    $derived_key = pbkdf1_sha1_base64 ( %params );


=head1 DESCRIPTION

PBKDF2 applies a pseudorandom function, such as hash-based message authentication
code (HMAC), to the input password or passphrase along with a salt value and
repeats the process many times to produce a derived key, which can then be used
as a cryptographic key in subsequent operations.


=head1 CONSTRUCTOR

=head2 Crypt::PBE::PBKDF1->new ( %params )

Params:

=over 4

=item * C<password> : The password to use for the derivation

=item * C<salt> : The salt to use for the derivation. This value should be generated randomly.

=item * C<hash> : Hash algorithm (default "sha1")

=item * C<count> : The number of internal iteractions to perform for the derivation key (default "1_000")

=back


=head1 METHODS

=head2 $pbkdf1->derived_key

Return the derived key in raw output (byte).

=head2 $pbkdf1->derived_key_base64

Return the derived key in Base64 encoded format.

=head2 $pbkdf1->derived_key_hex

Return the derived key in HEX format.

=head2 $pbkdf2->hash

Return the hash name.

=head2 $pbkdf2->count

Return the iteration count number.

=head2 $pbkdf2->derived_key_length

Return the derived key length.


=head1 FUNCTIONS

=head2 pbkdf1 ( hash => ..., password => ..., salt => ..., [ count => 1_000, dk_len => 0 ] )

Return derived key using PBKDF1 function:

    my $derived_key = pbkdf1 (
        hash     => 'sha1',
        password => 'mypassword',
        salt     => my_random_byte_sub(), 
        count    => 2_000
    );

    print length($derived_key)      # 20


=head2 pbkdf1_base64 ( hash => ..., password => ..., salt => ..., [ count => 1_000, dk_len => 0 ] )

Return derived key in Base64 using PBKDF1 function.

=head2 pbkdf1_hex ( hash => ..., password => ..., salt => ..., [ count => 1_000, dk_len => 0 ] )

Return derived key in HEX using PBKDF1 function.


=head2 EXPORTABLE HELPER FUNCTIONS

Return the derived key using MD2, MD5 and SHA1 digest:

=over 4

=item pbkdf1_md2

=item pbkdf1_md5

=item pbkdf1_sha1

=back

Return the derived key using MD2, MD5 and SHA1 digest in Base64:

=over 4

=item pbkdf1_md2_base64

=item pbkdf1_md5_base64

=item pbkdf1_sha1_base64

=back

Return the derived key using MD2, MD5 and SHA1 digest in HEX:

=over 4

=item pbkdf1_md2_hex

=item pbkdf1_md5_hex

=item pbkdf1_sha1_hex

=back


=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/giterlizzi/perl-Crypt-PBE/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/giterlizzi/perl-Crypt-PBE>

    git clone https://github.com/giterlizzi/perl-Crypt-PBE.git


=head1 AUTHOR

=over 4

=item * Giuseppe Di Terlizzi <gdt@cpan.org>

=back


=head1 SEE ALSO

=over 4

=item L<Crypt::PBE::PBKDF2>

=item [RFC2898] PKCS #5: Password-Based Cryptography Specification Version 2.0 (L<https://tools.ietf.org/html/rfc2898>)

=item [RFC8018] PKCS #5: Password-Based Cryptography Specification Version 2.1 (L<https://tools.ietf.org/html/rfc8018>)

=back


=head1 LICENSE AND COPYRIGHT

This software is copyright (c) 2020 by Giuseppe Di Terlizzi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Crypt/PBE.pm
package Crypt::PBE;

use strict;
use warnings;

use Carp;
use Crypt::PBE::PBES1;
use Crypt::PBE::PBES2;

our $VERSION = '0.100';

use Exporter qw(import);

my @JCE_PBE_ALGORITHMS = qw(

    PBEWithMD5AndDES

    PBEWithHmacSHA1AndAES_128
    PBEWithHmacSHA1AndAES_192
    PBEWithHmacSHA1AndAES_256

    PBEWithHmacSHA224AndAES_128
    PBEWithHmacSHA224AndAES_192
    PBEWithHmacSHA224AndAES_256

    PBEWithHmacSHA256AndAES_128
    PBEWithHmacSHA256AndAES_192
    PBEWithHmacSHA256AndAES_256

    PBEWithHmacSHA384AndAES_128
    PBEWithHmacSHA384AndAES_192
    PBEWithHmacSHA384AndAES_256

    PBEWithHmacSHA512AndAES_128
    PBEWithHmacSHA512AndAES_192
    PBEWithHmacSHA512AndAES_256
);

our @EXPORT_OK = @JCE_PBE_ALGORITHMS;

our %EXPORT_TAGS = ( 'jce' => \@JCE_PBE_ALGORITHMS );

# JCE algorithm PBEWith<digest>And<encryption>

my $pbes1_map = {
    'PBEWithMD2AndDES'  => { hash => 'md2',  encryption => 'des' },
    'PBEWithMD5AndDES'  => { hash => 'md5',  encryption => 'des' },
    'PBEWithSHA1AndDES' => { hash => 'sha1', encryption => 'des' },
};

my $pbes2_map = {
    'PBEWithHmacSHA1AndAES_128'   => { hmac => 'hmac-sha1',   encryption => 'aes-128' },
    'PBEWithHmacSHA1AndAES_192'   => { hmac => 'hmac-sha1',   encryption => 'aes-192' },
    'PBEWithHmacSHA1AndAES_256'   => { hmac => 'hmac-sha1',   encryption => 'aes-256' },
    'PBEWithHmacSHA224AndAES_128' => { hmac => 'hmac-sha224', encryption => 'aes-128' },
    'PBEWithHmacSHA224AndAES_192' => { hmac => 'hmac-sha224', encryption => 'aes-192' },
    'PBEWithHmacSHA224AndAES_256' => { hmac => 'hmac-sha224', encryption => 'aes-256' },
    'PBEWithHmacSHA256AndAES_128' => { hmac => 'hmac-sha256', encryption => 'aes-128' },
    'PBEWithHmacSHA256AndAES_192' => { hmac => 'hmac-sha256', encryption => 'aes-192' },
    'PBEWithHmacSHA256AndAES_256' => { hmac => 'hmac-sha256', encryption => 'aes-256' },
    'PBEWithHmacSHA384AndAES_128' => { hmac => 'hmac-sha384', encryption => 'aes-128' },
    'PBEWithHmacSHA384AndAES_192' => { hmac => 'hmac-sha384', encryption => 'aes-192' },
    'PBEWithHmacSHA384AndAES_256' => { hmac => 'hmac-sha384', encryption => 'aes-256' },
    'PBEWithHmacSHA512AndAES_128' => { hmac => 'hmac-sha512', encryption => 'aes-128' },
    'PBEWithHmacSHA512AndAES_192' => { hmac => 'hmac-sha512', encryption => 'aes-192' },
    'PBEWithHmacSHA512AndAES_256' => { hmac => 'hmac-sha512', encryption => 'aes-256' },
};

# PBES1 + PBDKF1

foreach ( keys %{$pbes1_map} ) {

    my $params   = $pbes1_map->{$_};
    my $sub_name = $_;

    no strict 'refs';    ## no critic

    *{$sub_name} = sub {
        my ( $password, $count ) = @_;
        my $pbes1 = Crypt::PBE::PBES1->new(
            password   => $password,
            count      => $count || 1_000,
            hash       => $params->{hash},
            encryption => $params->{encryption},
        );
        return $pbes1;
    };

}

# PBES2 + PBDKF2

foreach ( keys %{$pbes2_map} ) {

    my $params   = $pbes2_map->{$_};
    my $sub_name = $_;

    no strict 'refs';    ## no critic

    *{$sub_name} = sub {
        my ( $password, $count ) = @_;
        my $pbes2 = Crypt::PBE::PBES2->new(
            password   => $password,
            count      => $count || 1_000,
            hmac       => $params->{hmac},
            encryption => $params->{encryption},
        );
        return $pbes2;
    };

}

1;
__END__
=head1 NAME

Crypt::PBE - Perl extension for PKCS #5 Password-Based Encryption Algorithms

=head1 SYNOPSIS

    use Crypt::PBE qw(:jce);

    my $pbe = PBEWithMD5AndDES('mypassword');

    my $encrypted = $pbe->encrypt('secret'); # Base64 encrypted data

    print $pbe->decrypt($encrypted);


=head1 DESCRIPTION



=head2 PBES and PBKDF

=over 4

=item L<Crypt::PBE::PBKDF1> - Password-Based Key Derivation Function 1

=item L<Crypt::PBE::PBES1> - Password-Based Key Encryption Schema 1

=item L<Crypt::PBE::PBKDF2> - Password-Based Key Derivation Function 2

=item L<Crypt::PBE::PBES2> - Password-Based Key Encryption Schema 2

=back

=head2 EXPORTED JCE-STYLE FUNCTIONS

=head3 PBES1 (Password-Based Encryption Schema 1)

=over 4

=item * L<PBEWithMD2AndDES> : Password-Based Encryption with MD2 and DES

=item * L<PBEWithMD5AndDES> : Password-Based Encryption with MD5 and DES

=item * L<PBEWithSHA1AndDES> : Password-Based Encryption with SHA1 and DES

=back

=head3 PBES2 (Password-Based Encryption Schema 2)

=over 4

=item * L<PBEWithHmacSHA1AndAES_128> : Password-Based Encryption with SHA-1 HMAC and AES 128 bit

=item * L<PBEWithHmacSHA1AndAES_192> : Password-Based Encryption with SHA-1 HMAC and AES 192 bit

=item * L<PBEWithHmacSHA1AndAES_256> : Password-Based Encryption with SHA-1 HMAC and AES 256 bit

=item * L<PBEWithHmacSHA224AndAES_128> : Password-Based Encryption with SHA-224 HMAC and AES 128 bit

=item * L<PBEWithHmacSHA224AndAES_192> : Password-Based Encryption with SHA-224 HMAC and AES 192 bit

=item * L<PBEWithHmacSHA224AndAES_256> : Password-Based Encryption with SHA-224 HMAC and AES 256 bit

=item * L<PBEWithHmacSHA256AndAES_128> : Password-Based Encryption with SHA-256 HMAC and AES 128 bit

=item * L<PBEWithHmacSHA256AndAES_192> : Password-Based Encryption with SHA-256 HMAC and AES 192 bit

=item * L<PBEWithHmacSHA256AndAES_256> : Password-Based Encryption with SHA-256 HMAC and AES 256 bit

=item * L<PBEWithHmacSHA384AndAES_128> : Password-Based Encryption with SHA-384 HMAC and AES 128 bit

=item * L<PBEWithHmacSHA384AndAES_192> : Password-Based Encryption with SHA-384 HMAC and AES 192 bit

=item * L<PBEWithHmacSHA384AndAES_256> : Password-Based Encryption with SHA-384 HMAC and AES 256 bit

=item * L<PBEWithHmacSHA512AndAES_128> : Password-Based Encryption with SHA-512 HMAC and AES 128 bit

=item * L<PBEWithHmacSHA512AndAES_192> : Password-Based Encryption with SHA-512 HMAC and AES 192 bit

=item * L<PBEWithHmacSHA512AndAES_256> : Password-Based Encryption with SHA-512 HMAC and AES 256 bit

=back

=head1 SEE ALSO

=over 4

=item [RFC2898] PKCS #5: Password-Based Cryptography Specification Version 2.0 (L<https://tools.ietf.org/html/rfc2898>)

=item [RFC8018] PKCS #5: Password-Based Cryptography Specification Version 2.1 (L<https://tools.ietf.org/html/rfc8018>)

=item [RFC6070] PKCS #5: Password-Based Key Derivation Function 2 (PBKDF2) - Test Vectors (L<https://tools.ietf.org/html/rfc6070>)

=item [RFC2307] An Approach for Using LDAP as a Network Information Service (L<https://tools.ietf.org/html/rfc2307>)

=back


=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/giterlizzi/perl-Crypt-PBE/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/giterlizzi/perl-Crypt-PBE>

    git clone https://github.com/giterlizzi/perl-Crypt-PBE.git


=head1 AUTHOR

=over 4

=item * Giuseppe Di Terlizzi <gdt@cpan.org>

=back


=head1 LICENSE AND COPYRIGHT

This software is copyright (c) 2020 by Giuseppe Di Terlizzi.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
